.github/workflows/inspektor-gadget.yml

```diff
@@ -1342,6 +1342,55 @@ jobs:
           BUILDER_IMAGE=${{ needs.build-helper-images.outputs.ebpf_builder_image }} \
           push-gadgets -o install/ig
 
+  test-gadgets:
+    name: Test gadgets
+    #level 2
+    needs:
+      - build-ig
+      - build-helper-images
+    runs-on: ubuntu-latest
+    permissions:
+      # allow publishing container image
+      # in case of public fork repo/packages permissions will always be read
+      contents: read
+      packages: write
```

Comment:
CI passes only when we provide the write permission, even though we are not publishing the images to the remote repository, we are building them locally, I would like to understand why this is required.

Reply:
https://github.com/pawarpranav83/inspektor-gadget/actions/runs/8575164865/job/23503783080

Reply:
Yeah, you are right, the test doesn't work when directly specifying the value of the gadget repo and tag.
https://github.com/pawarpranav83/inspektor-gadget/actions/runs/8601351504/job/23568643301

I think the gadget-image doesn't exist, but still its running the test, and no output is captured.
Will try to test this.

Reply:
Error: 
```
~/Desktop/Repos/inspektor-gadget/gadgets$ make test
CGO_ENABLED=0 \
IG_EXPERIMENTAL=true \
IG=ig \
go test -exec 'sudo -E' -v ./...
?       github.com/inspektor-gadget/inspektor-gadget/gadgets/testing    [no test files]
=== RUN   TestTraceOpen
    command.go:113: Start command(Run_trace_open)
    docker.go:117: Container "test-trace-open" output:
    command.go:130: Stop command(Run_trace_open)
    command.go:132: Command returned(Run_trace_open):
        time="2024-04-08T19:58:08+05:30" level=info msg="Experimental features enabled"
        
        
    match.go:71: output doesn't contain the expected entry
        captured:
        expected:
        {"runtime":{"runtimeName":"docker","containerName":"test-trace-open"},"k8s":{},"mountnsid":0,"pid":0,"uid":1000,"gid":1111,"comm":"cat","ret":3,"flags":0,"mode":0,"fname":"/dev/null"}
--- FAIL: TestTraceOpen (15.69s)
FAIL
FAIL    github.com/inspektor-gadget/inspektor-gadget/gadgets/trace_open/test    15.732s
FAIL
make: *** [Makefile:57: test] Error 1
```
Removed the image and skipped the build test, with no image the test is still running, and no output is displayed.

Reply:
But if that image doesn't exist it does give an error, so its not about image not existing ig: 
```
~/Desktop/Repos/inspektor-gadget/gadgets$ make test
CGO_ENABLED=0 \
IG_EXPERIMENTAL=true \
IG=ig \
go test -exec 'sudo -E' -v ./...
?       github.com/inspektor-gadget/inspektor-gadget/gadgets/testing    [no test files]
=== RUN   TestTraceOpen
    command.go:113: Start command(Run_nice)
    docker.go:117: Container "test-trace-open" output:
    command.go:130: Stop command(Run_nice)
    command.go:132: Command returned(Run_nice):
        time="2024-04-08T20:18:25+05:30" level=info msg="Experimental features enabled"
        Error: getting gadget info: getting gadget image: ensuring image presence: pulling image (if missing) "nice": downloading to local repository: failed to resolve ghcr.io/inspektor-gadget/gadget/nice:latest: GET "https://ghcr.io/v2/inspektor-gadget/gadget/nice/manifests/latest": GET "https://ghcr.io/token?scope=repository%3Ainspektor-gadget%2Fgadget%2Fnice%3Apull&service=ghcr.io": response status code 403: denied: requested access to the resource is denied
        
        
    command.go:134: 
                Error Trace:    /home/jinx83/Desktop/Repos/inspektor-gadget/pkg/testing/command/command.go:134
                                                        /home/jinx83/Desktop/Repos/inspektor-gadget/pkg/testing/teststeps.go:78
                                                        /home/jinx83/Desktop/Repos/inspektor-gadget/pkg/testing/teststeps.go:92
                                                        /home/jinx83/Desktop/Repos/inspektor-gadget/gadgets/trace_open/test/trace_open_test.go:92
                Error:          Received unexpected error:
                                exit status 1
                Test:           TestTraceOpen
                Messages:       failed to kill command(Run_nice)
--- FAIL: TestTraceOpen (13.78s)
FAIL
FAIL    github.com/inspektor-gadget/inspektor-gadget/gadgets/trace_open/test    13.822s
FAIL
make: *** [Makefile:57: test] Error 1
```

Reply:
The gadget is working fine, but since we are using StartAndStop, the command is started and when no gadget is found, it pulls the gadget from the remote repo, but meanwhile we run the container command till the gadget is run that command is missed so therefore, it shows no output.
But it still doesn't answer why it is pulling the image if we are building it locally.

Reply:
Since the default value of `GADGET_REPOSITORY=ghcr.io/inspektor-gadget/gadget` and `GADGET_TAG=latest`, in case of local CI repo value was `ghcr.io/pawarpranav83/gadget` and tag value as `temp-fw` instead, which led to pulling the image from remote repository since we don't pass these env vars to make test (we rely on the feature of ig instead where it adds the prefix and suffix to gadget name automatically, and these are the default values).

Solution: We can manually set the image name with GADGET_REPOSITORY and GADGET_TAG environment variables, but since in the main branch CI, images are tagged with default values, it would use the local gadget-images.

Reply:
Thanks for describing the issue with the `GADGET_REPOSITORY` and `GADGET_TAG`, it's clear now.

I agree we should pass the `GADGET_REPOSITORY` and `GADGET_TAG` values to the test and avoid letting IG add the default repo and tag (Sorry, I think I suggested doing that). 

Now, as I mentioned [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/2607#discussion_r1545860499), I still consider the `test-gadgets` job shouldn't build the gadgets again but use the ones built and pushed by the `build-and-push-gadgets` job. 

On the other hand, you can solve the issue of IG not capturing the event because it's busy pulling the image by repeating the command many times. Notice that in the current tests, we use `BusyboxPodRepeatCommand` which executes the command periodically every `0.1` seconds. So, here you could do:

```go
containerFactory.NewContainer(cn, "while true; do setuidgid 1000:1111 cat /dev/null; sleep 0.1; done"),
```

Probably, we should implement the `WithRepeatedCmd(0.1)` option for `NewContainer`. But that's something we can do in another PR.

Reply:
I think such a helper could help. Btw, to use the while loop as you suggested, it's also needed to pass `containers.WithStartStop()`, otherwise the tests keeps running forever.